standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DirectSqlUpdatePart.java

```diff
@@ -892,18 +892,25 @@ public List<Void> run(List<Long> input) throws Exception {
         sdIdToNewCdId.containsKey(sdId) ? sdIdToNewCdId.get(sdId) : cdId);
     updateSDInBatch(validSdIds, idToSd, sdIdToCdId);
 
-    List<Long> usedIds = Batchable.runBatched(maxBatchSize, cdIdsMayDelete,
+    Set<Long> usedIds = new HashSet<>(Batchable.runBatched(maxBatchSize, cdIdsMayDelete,
         new Batchable<Long, Long>() {
           @Override
           public List<Long> run(List<Long> input) throws Exception {
             String idLists = MetaStoreDirectSql.getIdListForIn(input);
-            String queryText = "select \"CD_ID\" from \"SDS\" where \"CD_ID\" in ( " + idLists + ")";
+            String queryText = "select \"CD_ID\" from \"SDS\" where \"CD_ID\" in ( "
```

Comment:
why not `select distinct`

Reply:
https://www.linkedin.com/pulse/optimizing-sql-performance-distinct-vs-group-fact-analysis-rahman-z5sec/
For the relational DBs `SELECT DISTINCT`  is typically faster or at least simpler than `GROUP BY`

Reply:
Done